dropbox/dropbox.py

```diff
@@ -299,6 +331,46 @@ def request(self,
         else:
             return deserialized_result
 
+    def _check_and_refresh_access_token(self):
+        """
+        Checks if access token needs to be refreshed and refreshes if possible
+
+        :return:
+        """
+        can_refresh = self._oauth2_refresh_token and self._app_key and self._app_secret
+        needs_refresh = self._oauth2_access_token_expiration and \
+                        (datetime.utcnow() + timedelta(seconds=TOKEN_EXPIRATION_BUFFER)) >= \
+                        self._oauth2_access_token_expiration
+        needs_token = not self._oauth2_access_token
+        if (needs_refresh or needs_token) and can_refresh:
+            self._refresh_access_token()
+
+    def _refresh_access_token(self):
+        """
+        Refreshes an access token via refresh token if available
+
+        :return:
+        """
+
+        if not (self._oauth2_refresh_token and self._app_key and self._app_secret):
+            self._logger.warning('Unable to refresh access token without refresh token, app key, and app secret')
+            return
+
+        self._logger.info('Refreshing access token.')
+        url = "https://api.dropboxapi.com/oauth2/token"
+        body = {'grant_type': 'refresh_token',
+                  'refresh_token': self._oauth2_refresh_token,
+                  'client_id': self._app_key,
+                  'client_secret': self._app_secret,
+                }
+
+        res = self._session.post(url, data=body)
+        res.raise_for_status()
```

Comment:
Why would we want to not fail on failure to refresh token?  Given this only affects short lived tokens, it wont change any existing flows.  I think that throwing expired_access_token seems incorrect because the issue is that refreshing failed not just that the token expired

Reply:
To be more specific, we don't really want to throw error code `expired_access_token`. We just need to find away to throw `invalid_access_token` to tell the app to log_out their user. That's why we can't directly do `raise_for_status` here